### PR TITLE
[RFC] cleanup bootargs for ARMv7 architectures

### PR DESCRIPTION
**Pull request by etienne-lms**
!!! This is a Request For Comment !!!   **(updated: fix typos)**

This change may affect existing platforms. Need to synchronize with the maintainers.

_- sumup -_

This change aims at rationalizing how ARMv7 generic entry handles the optee boot arguments and standard (linux) boot arguments.

Some platform uses `CFG_TEE_GDB_BOOT` when nonsecure entry point (after optee init) is the dear linux-kernel. `CFG_TEE_GDB_BOOT` relays boot arguments (cpu registers r0/r1/r2/r3) from optee entry to nonsecure (linux) entry. Current implementation scratches DeviceTree address boot argument.

Some platform uses `CFG_TEE_GDB_BOOT` together with `CFG_BUILT_IN_ARGS`. Optee boot arguments are hardcoded while linux boot arguments are relayed to kernel entry.

Optee and linux boot arguments are compatible: r2=dt, r1=unused-in-optee, r0=unused-in-linux. Code could mutualize: get optee boot args and relay linux boot args. 

_ - benefit -_

Real effective benefit of this change is to ARMv7 optee to access DT and share it with a booted nonsecure linux kernel. This is not possible with current code unless using hard coded boot arguments.

This change tries to simplify boot arguments configuration for optee users rather than only fixing this DT address issue.

_- impacted config -_

`CFG_TEE_GDB_BOOT` is deprecated (removed). It is replaced with `CFG_RELAY_STD_BOOTARGS`.
`CFG_BUILT_IN_ARGS` is deprecated (removed). 
`CFG_PAGEABLE_ADDR,` `CFG_DT_ADDR` and `CFG_NS_ENTRY_ADDR` are preserved.
Platform can define independently each of these in case hardcoded boot arguments are required.

### CODE REVIEW

**Comment by etienne-lms** [comment]
@MrVan, @yanyan-wrs, could you have a look at this ?

regards,
etienne

**Comment by yanyan-wrs** [comment]
I agree with this change. The `CFG_TEE_GDB_BOOT` is confusing, and does not reflect its real purpose, and `CFG_RELAY_STD_BOOTARGS` is what we expect.
And without the ARM Trusted Firmware, the least intrusive method of booting OP-TEE is to make it as transparent as possible; one way is to insert it between the generic bootstrap and the Linux payload. There should be as less change as possible to the bootstrap's interface, i.e., it can keep passing the same arguments to OP-TEE as it does with Linux. And, as OP-TEE requires extra arguments, which we do not expect to get from a generic bootstrap, the only reasonable way is to use some hard-coded defines.

**Comment by MrVan** [comment]
Happy to see the changes. I did a basic test on i.MX6Q-SDB, it works well.

File: mk/config.mk [hunk 1]
```diff
@@ -197,15 +197,5 @@ CFG_TEE_CORE_EMBED_INTERNAL_TESTS ?= y
 # with specific core number and non-secure entry address.
 CFG_BOOT_SECONDARY_REQUEST ?= n
 
-# For firmwares that do not pass arguments to OP-TEE, this option enables the
-# built-in arguments which are specified as CFG items in the build time:
-# CFG_PAGEABLE_ADDR: pageable data address, normally passed in R0;
-# CFG_NS_ENTRY_ADDR: NS World entry address, normally passed in LR;
-# CFG_DT_ADDR: Device Tree data address, normally passed in R2;
-CFG_BUILT_IN_ARGS ?= n
-CFG_PAGEABLE_ADDR ?= 0x0
```
Comment (jenswi-linaro) [inline mk/config.mk]: Please document these three variables here and keep the commented out.

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 2]
```diff
@@ -156,26 +149,19 @@ UNWIND(	.fnstart)
...
+	ldr	r5, =CFG_NS_ENTRY_ADDR
+#endif
+#if defined(CFG_PAGEABLE_ADDR)
+	ldr	r4, =CFG_PAGEABLE_ADDR
 #endif
+#if defined(CFG_DT_ADDR)
+	ldr	r6, =CFG_DT_ADDR
+#endif
+#if defined(CFG_RELAY_STD_BOOTARGS)
+	mov	sp, r1		/* Save Boot argument (std bootarg #1) */
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: Why not put this in `r7` directly?

File: core/arch/arm/kernel/generic_entry_a32.S [hunk 3]
```diff
@@ -156,26 +149,19 @@ UNWIND(	.fnstart)
...
 
-#if !defined(CFG_BUILT_IN_ARGS) && (defined(CFG_WITH_PAGER) || defined (CFG_DT))
-	/*
-	 * r0/r2 are used for different purposes by GDB boot and pager/DT
-	 * so these configurations are incompatible. however, if built-in
-	 * argumements are used, r0/r2 are free for GDB boot use.
-	 */
-#error CFG_TEE_GDB_BOOT and (CFG_WITH_PAGER || CFG_DT) are incompatible
+#if defined(CFG_NS_ENTRY_ADDR)
+	ldr	r5, =CFG_NS_ENTRY_ADDR
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: I would prefer something like (same principle for the other optionally hard coded addresses):
```
#if defined(CFG_NS_ENTRY_ADDR)
    ldr   r5, =CFG_NS_ENTRY_ADDR
#else
    mov   r5, lr
#endif
```

Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: ok

File: core/arch/arm/kernel/generic_entry_a32.S (hunk 2 above)
Comment (etienne-lms) [inline core/arch/arm/kernel/generic_entry_a32.S]: this is the ugly part.
`r7` is scratched by the memory moves in `reset_primary` (pager + kasan inits).
i did not want to use the stack before kasan was setup, no no push to stack before

File: mk/config.mk (hunk 1 above)
Comment (etienne-lms) [inline mk/config.mk]: ok. i just don't want them to be default set.

File: core/arch/arm/kernel/generic_entry_a32.S (hunk 2 above)
Comment (jenswi-linaro) [inline core/arch/arm/kernel/generic_entry_a32.S]: OK, I'd rather not scratch `r7` in `copy_init` then. The performance decrease is probably not noticeable. And if it's truly a problem we could use `sp` as a scratch registers in `copy_init` instead.

File: core/arch/arm/arm.mk [hunk 4]
```diff
@@ -40,6 +40,15 @@ $(error Error: CFG_CORE_SANITIZE_KADDRESS not compatible with CFG_WITH_PAGER)
 endif
 endif
 
+ifeq ($(CFG_ARM32_core),y)
+# Configration directive reated to ARMv7 optee boot: boot arguments:
```
Comment (jbech-linaro) [inline core/arch/arm/arm.mk]: s/reated/related/

**Comment by jenswi-linaro** [comment]
Why do we need to have `CFG_RELAY_STD_BOOTARGS`? If `CFG_WITH_ARM_TRUSTED_FW` isn't defined we could do what `CFG_RELAY_STD_BOOTARGS == y` would have done unconditionally. It would make it slightly less complicated at almost no cost.

**Comment by etienne-lms** [comment]
> Why do we need to have CFG_RELAY_STD_BOOTARGS? If CFG_WITH_ARM_TRUSTED_FW isn't defined

if a platform want to boot a non linux and don't want to leak data from secure boot.
probably useless...

File: core/arch/arm/arm.mk (hunk 4 above)
Comment (etienne-lms) [inline core/arch/arm/arm.mk]: thanks.

**Comment by jenswi-linaro** [comment]
Let's deal with that problem when it's more than a hypothetical problem.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by etienne-lms** [comment]
squashed, tags applied.

**Comment by jbech-linaro** [comment]
Checkpatch found a spelling error on the second commit.

**Comment by etienne-lms** [comment]
typo fixed. i should run the nice checkpatch script before pushing!

**Comment by jforissier** [comment]
Hi @etienne-lms, could you please rebase?

**Comment by etienne-lms** [comment]
sure. i'll run another nonreg test to be shure there is no hidden issue....

**Comment by etienne-lms** [comment]
rebase done.

**Pull request merged**